fab/sampling_methods/transition_operators/hmc.py

```diff
@@ -102,19 +102,27 @@ def get_epsilon(self, i: int, n: int) -> torch.Tensor:
     def joint_log_prob(self, point: Point, p, mass_matrix, U):
         return - U(point) - self.kinetic_energy(p, mass_matrix)
 
-    def metropolis_acceptance_prob(self, point_proposed: Point, point_current: Point,
-                                   p_proposed: torch.Tensor, p_current: torch.Tensor,
-                                   mass_matrix: torch.Tensor, U: Callable):
+    def metropolis_accept(self, point_proposed: Point, point_current: Point,
+                          p_proposed: torch.Tensor, p_current: torch.Tensor,
+                          mass_matrix: torch.Tensor, U: Callable):
         log_prob_current = self.joint_log_prob(point_current, p_current, mass_matrix, U)
         log_prob_proposed = self.joint_log_prob(point_proposed, p_proposed, mass_matrix, U)
-        acceptance_prob = torch.exp(log_prob_proposed - log_prob_current)
-        # reject samples with nan acceptance probability
-        acceptance_probability = torch.nan_to_num(acceptance_prob,
-                                                  nan=0.0,
-                                                  posinf=0.0,
-                                                  neginf=0.0)
-        acceptance_prob = torch.clamp(acceptance_probability, min=0.0, max=1.0)
-        return acceptance_prob.detach()
+        with torch.no_grad():
+            log_acceptance_prob = log_prob_proposed - log_prob_current
+            # reject samples with nan acceptance probability
+            valid_samples = torch.isfinite(log_acceptance_prob)
+            log_acceptance_prob = torch.nan_to_num(log_acceptance_prob,
+                                                      nan=-1e30,
+                                                      posinf=-1e30,
+                                                      neginf=-1e30)
+            accept = log_acceptance_prob > torch.log(torch.rand(
```

Comment:
@THargreaves is this what you meant?

Reply:
Basically there. I think you can make it more stable by noting that if X ~ Unif(0, 1) then -log(X) ~ exp(1) (note the negative). That way you don't need to take the log of the uniform but can rather compare directly to an exponential sample.